python/ray/experimental/signal.py

```diff
@@ -0,0 +1,168 @@
+
+def _get_source_key(source_id):
+    if  type(source_id) is ray.actor.ActorHandle:
+        source_id_key = source_id._actor_id.id()
+    else:
+        source_id_key = source_id.id()
+    return source_id_key
+
+def send(signal, source_id = None):
+    """Send signal on behalf of source_id.
+
+    Each signal is identified by (source_id, index), where index is incremented
+    every time a signal is sent, starting from 1. Receiving this signal,
+    requires waiting on (source_id, index).
+
+    Args:
+        signal: signal to be sent.
+        source_id: If empty, initialize to the id of the task/actor
+                   invoking this function.
+    """
+    if source_id == None:
+        actors = ray.worker.global_worker.actors
+        if len(actors) > 0:
+            source_key, = actors.keys()
+        else:
+            # no actors; this function must have been called from a task
+            source_key = ray.worker.global_worker.current_task_id.id()
+    else:
+        source_key = source_id.id()
+    # get index for this signal.
+    index = ray.worker.global_worker.redis_client.incr(source_key)
+    object_id = _get_signal_object_id_from_binary(source_key, index)
+    ray.worker.global_worker.store_and_register(object_id, signal)
+
+def receive(source_ids, timeout=float('inf')):
+    """Get all signals from each source in source_ids.
+
+    For each source_id in source_ids, this function returns all signals
+    generated by (or on behalf of) source_id since the last receive() or
+    forget() were invoked on source_id. If this is the first call on
+    source_id, this function returns all signals generated by (or on
+    behalf of) source_id so far.
+
+    Args:
+        source_ids: list of source ids whose signals are returned.
+        timeout: time it receives for new signals to be generated. If none,
+                 return when timeout experies. Measured in seconds.
+
+    Returns:
+        The list of signals generated for each source in source_ids. They
+        are returned as a list of pairs (source_id, signal). If there are
+        more signals for the same source_id, they are inserted in the list
+        in the order they were generated.
+    """
+    if not hasattr(ray.worker.global_worker, "signal_counters"):
+        ray.worker.global_worker.signal_counters = dict()
+
+    signal_counters = ray.worker.global_worker.signal_counters
+    results = []
+    previous_time = time.time()
+    remaining_time = timeout
+
+    for source_id in source_ids:
+        if not source_id in signal_counters:
+            signal_counters[source_id] = 1
+
+    for source_id in source_ids:
+        while True:
+            object_id = _get_signal_object_id(source_id, signal_counters[source_id])
+            ready_ids, _ = ray.wait([object_id], num_returns=1, timeout=0)
+            if len(ready_ids) > 0:
+                results.append((source_id, ray.get(object_id)))
```

Comment:
Here, we should check whether the return value of `ray.get(object_id)` is an instance of `Signal`. You can do that like this:
```
value = ray.get(object_id)
if isinstance(value, Signal):
  # Process the signal.
```

Reply:
When there is a node failure and an actor dies, an exception will be thrown when you try to `ray.get` the object. You can handle it like this:
```
try:
  ray.get(object_id)
catch:
  # The actor is dead.
```

Note that this will not work for tasks since the backend will try to reconstruct the task. Also, this might not work right now because of #3715.